File: 39-combination-sum/ken/39.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        ans = []
        dp = {}

        def sameList(a, b):
            a.sort()
            b.sort()

            return a == b

        def appendIfNot(l, e):
            for i in l:
                if sameList(i, e):
                    return
            l.append(e)

        def rec(target):
            if target in dp:
                return dp[target]

            ans = []
            for candidate in candidates:
                nextTarget = target - candidate
                if nextTarget == 0:
                    appendIfNot(ans, [candidate])
                elif nextTarget > 0:
                    subCombinations = rec(nextTarget)

                    if len(subCombinations) == 0:
                        continue

                    for subCombination in subCombinations:
                        cc = list(subCombination)
                        cc.append(candidate)
                        appendIfNot(ans, cc)

            if len(ans) > 0:
                dp[target] = ans

            return ans

        ans = rec(target)

        return ans

File: 39-combination-sum/leochang/39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
		vector<int> cand;
		sort(candidates.begin(), candidates.end());
        if (candidates.size() == 0) return ans;
        dfs(candidates, target, ans, cand, 0);
        return ans;
    }
private:
    void dfs(vector<int>& candidates, int target, vector<vector<int>> &ans, vector<int> &cand, int start) {
        if (0 == target) {
            ans.push_back(cand);
            return;
        }
        for (int i = start; i < candidates.size() && candidates[i] <= target; i++) {
            cand.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], ans, cand, i);
            cand.pop_back();
        }
    }
};
